lib/glueby/contract/token_tx_builder.rb

```diff
@@ -0,0 +1,201 @@
+        tx.outputs << Tapyrus::TxOut.new(value: amount, script_pubkey: receiver_colored_script)
+
+        fill_change_token(tx, sender, sum_token - amount, color_id)
+        fill_change_tpc(tx, sender, sum_tpc - fee)
+        sender.internal_wallet.sign_tx(tx)
+      end
+
+      def create_burn_tx(color_id:, sender:, amount: 0, fee_provider: FixedFeeProvider.new)
+        tx = Tapyrus::Tx.new
+        fee = fee_provider.fee(tx)
+
+        utxos = sender.internal_wallet.list_unspent
+        dust = 600 # in case that the wallet has output which has just fee amount.
+        sum_tpc, outputs = collect_uncolored_outputs(utxos, fee + dust)
+        fill_input(tx, outputs)
+
+        sum_token, outputs = collect_colored_outputs(utxos, color_id, amount)
+        fill_input(tx, outputs)
+
+        fill_change_token(tx, sender, sum_token - amount, color_id) if amount.positive?
+
+        fill_change_tpc(tx, sender, sum_tpc - fee)
+        sender.internal_wallet.sign_tx(tx)
+      end
+
+      def fill_input(tx, outputs)
+        outputs.each do |output|
+          out_point = Tapyrus::OutPoint.new(output[:txid].rhex, output[:vout])
+          tx.inputs << Tapyrus::TxIn.new(out_point: out_point)
+        end
+      end
+
+      def fill_change_tpc(tx, wallet, change)
+        return unless change.positive?
+
+        change_script = Tapyrus::Script.parse_from_addr(wallet.internal_wallet.change_address)
+        tx.outputs << Tapyrus::TxOut.new(value: change, script_pubkey: change_script)
+      end
+
+      def fill_change_token(tx, wallet, change, color_id)
+        return unless change.positive?
+
+        change_script = Tapyrus::Script.parse_from_addr(wallet.internal_wallet.change_address)
+        change_colored_script = change_script.add_color(color_id)
+        tx.outputs << Tapyrus::TxOut.new(value: change, script_pubkey: change_colored_script)
+      end
+
+      def collect_uncolored_outputs(results, amount)
+        results.inject([0, []]) do |sum, output|
+          script = Tapyrus::Script.parse_from_payload(output[:script_pubkey].htb)
+          next sum if script.cp2pkh? || script.cp2sh?
+
+          new_sum = sum[0] + output[:amount]
+          new_outputs = sum[1] << output
+          return [new_sum, new_outputs] if new_sum >= amount
+
+          [new_sum, new_outputs]
+        end
+        raise Glueby::Contract::Errors::InsufficientFunds
+      end
+
+      def collect_colored_outputs(results, color_id, amount = 0)
+        results = results.inject([0, []]) do |sum, output|
+          script = Tapyrus::Script.parse_from_payload(output[:script_pubkey].htb)
+          next sum unless script.cp2pkh? || script.cp2sh?
+          next sum unless Tapyrus::Color::ColorIdentifier.parse_from_payload(script.chunks[0].pushed_data) == color_id
+
+          new_sum = sum[0] + output[:amount]
+          new_outputs = sum[1] << output
+          return [new_sum, new_outputs] if new_sum >= amount && amount.positive?
```

Comment:
Can the amount be negative?

Reply:
if specified amount is 0, this method return all utxos with the specified color_id.
This option will be used when we burn all token in a wallet.
I added the comment about it.(5ae8537)